### PR TITLE
Calculate multiple metrics

### PR DESCRIPTION
**Pull request by rabsr**
This change will provide users an option scoring_functions to calculate list of metrics for each pipeline. The calculated metrics will be available from attribute cv_results_.
This will provide users with more insights for each pipeline executed.

Usage:
```
automl = autosklearn.classification.AutoSklearnClassifier(
    time_left_for_this_task=30,
    per_run_time_limit=10,
    scoring_functions=['accuracy', 'balanced_accuracy', 'log_loss'],)
automl.fit(X_train,y_train)
print(automl.cv_results_)
```

### CODE REVIEW

**Comment by mfeurer** [comment]
This implements #981 I guess?

File: autosklearn/metrics/__init__.py [hunk 1]
```diff
@@ -341,19 +341,22 @@ def calculate_score(
     prediction: np.ndarray,
     task_type: int,
     metric: Scorer,
-    all_scoring_functions: bool = False
+    scoring_functions: list = None
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:344]: This should be something like `scoring_functions: List[Scorer] = None`, where `Scorer` is the scorer class in `autosklearn.metric`.

File: autosklearn/metrics/__init__.py [hunk 2]
```diff
@@ -341,19 +341,22 @@ def calculate_score(
...
-    if all_scoring_functions:
+    if scoring_functions:
         score_dict = dict()
         if task_type in REGRESSION_TASKS:
             # TODO put this into the regression metric itself
             cprediction = sanitize_array(prediction)
             metric_dict = copy.copy(REGRESSION_METRICS)
             metric_dict[metric.name] = metric
             for metric_ in REGRESSION_METRICS:
+                if metric_ not in scoring_functions:
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:357]: Wouldn't it make more sense to loop through the scoring functions and apply all which are in there? This would allow custom scoring functions, while this change doesn't.

File: autosklearn/metrics/__init__.py [hunk 3]
```diff
@@ -369,6 +372,9 @@ def calculate_score(
             metric_dict = copy.copy(CLASSIFICATION_METRICS)
             metric_dict[metric.name] = metric
             for metric_ in metric_dict:
+                if metric_ not in scoring_functions:
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:375]: Same here.

File: autosklearn/metrics/__init__.py [hunk 4]
```diff
@@ -388,14 +394,20 @@ def calculate_score(
                         continue
                     else:
                         raise e
+
+        if metric and metric.name not in score_dict.keys():
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:398]: Together with the proposed change above, this should become redundant, right?

File: test/test_evaluation/test_train_evaluator.py [hunk 5]
```diff
@@ -557,12 +557,16 @@ def test_file_output(self, loss_mock):
         configuration = unittest.mock.Mock(spec=Configuration)
         queue_ = multiprocessing.Queue()
         loss_mock.return_value = None
+        scorer_list = ['accuracy', 'balanced_accuracy', 'f1_macro', 'f1_micro',
```
Comment (mfeurer) [inline test/test_evaluation/test_train_evaluator.py:560]: Could you please extract this so that it can be reused by multiple tests?

**Review by mfeurer — COMMENTED** [review]
Thanks a lot for the PR! I left some initial comments you could address, and then it would be good to incrementally work on the following things:
1. rebase on the development branch to fix the current issues CI
2. add an example about tracking auxiliary metrics
3. think about additional unit tests to test this new functionality

Looking forward to the changes

File: autosklearn/metrics/__init__.py (hunk 1 above)
Comment (rabsr) [reply autosklearn/metrics/__init__.py]: Made necessary changes.

File: autosklearn/metrics/__init__.py (hunk 2 above)
Comment (rabsr) [reply autosklearn/metrics/__init__.py]: Makes sense. Made required changes.

File: autosklearn/metrics/__init__.py (hunk 3 above)
Comment (rabsr) [reply autosklearn/metrics/__init__.py]: Done

File: test/test_evaluation/test_train_evaluator.py (hunk 5 above)
Comment (rabsr) [reply test/test_evaluation/test_train_evaluator.py]: Refactored code.

File: autosklearn/metrics/__init__.py (hunk 4 above)
Comment (rabsr) [reply autosklearn/metrics/__init__.py:387]: This change is added to make sure metric passed is calculated. It may possible that the list of scorers does not contain metric scorer required for train/test scoring.

**Comment by rabsr** [comment]
@mfeurer 

> This implements #981 I guess?

Yeah. This PR implements #981 

Made changes based on review comments and rebased with latest development. 

Todo

- [ ] add an example about tracking auxiliary metrics 
- [ ] additional unit tests to test this new functionality

I will be working on these items.

File: autosklearn/metrics/__init__.py (hunk 4 above)
Comment (mfeurer) [reply autosklearn/metrics/__init__.py:387]: Got it, thanks a lot. Another small question, can `metric` really be `None`?

File: test/test_evaluation/test_train_evaluator.py [hunk 6]
```diff
@@ -546,12 +546,13 @@ def test_file_output(self, loss_mock):
...
         queue_ = multiprocessing.Queue()
         loss_mock.return_value = None
+        scorer_list = SCORER_LIST
 
         evaluator = TrainEvaluator(self.backend_mock, queue=queue_,
                                    configuration=configuration,
                                    resampling_strategy='cv',
                                    resampling_strategy_args={'folds': 5},
-                                   all_scoring_functions=True,
+                                   scoring_functions=scorer_list,
```
Comment (mfeurer) [inline test/test_evaluation/test_train_evaluator.py:555]: What's the reason to assign the constant to a local variable and use the local variable here?

Comment (rabsr) [reply test/test_evaluation/test_train_evaluator.py]: Missed it. Made changes.

File: autosklearn/estimators.py [hunk 7]
```diff
@@ -42,7 +43,8 @@ def __init__(
         logging_config=None,
         metadata_directory=None,
         metric=None,
-        scoring_functions=None,
+        scoring_functions: List[Scorer] = None,
```
Comment (mfeurer) [inline autosklearn/estimators.py:46]: Sorry for not being clear enough earlier, could you please make this `Optional[List[Scorer]]` to explicitly state that this can be `None`?

File: autosklearn/metrics/__init__.py [hunk 8]
```diff
@@ -341,7 +341,7 @@ def calculate_score(
     prediction: np.ndarray,
     task_type: int,
     metric: Scorer,
-    scoring_functions: list = None
+    scoring_functions: List[Scorer] = None
```
Comment (mfeurer) [inline autosklearn/metrics/__init__.py:344]: Please also add the `Optional` here.

**Comment by mfeurer** [comment]
This looks great, looking forward to the example and unit tests.

File: autosklearn/estimators.py (hunk 7 above)
Comment (rabsr) [reply autosklearn/estimators.py]: Done

File: autosklearn/metrics/__init__.py (hunk 8 above)
Comment (rabsr) [reply autosklearn/metrics/__init__.py]: Done

File: autosklearn/metrics/__init__.py (hunk 4 above)
Comment (rabsr) [reply autosklearn/metrics/__init__.py]: Metric cant be None. Removed check.

**Review by mfeurer — COMMENTED** [review]
A few updates from my side:
* I cannot reproduce the failure for the pandas example locally, so I think everything is fine here. Nevertheless, I restarted the tests on travis-ci.
* The code looks good and I would merge it as is, therefore, you could now move to adding unit tests (I've been thinking about one for the score calculation function, and making the tests in `test_train_evaluator` stricter by checking the entries in the additional run info, please let me know if you have any questions regarding that) and an example on how to use this new feature.

I'm really looking forward to this new feature as I actually also just found a good use case within our research :)

**Comment by rabsr** [comment]
@mfeurer Test cases and example will take some time.   Currently  I am on vacation for 2 weeks. I will start working on test cases and example once I am back. 
> I actually also just found a good use case within our research 
If possible, can you highlight what is the use case you aee looking forward to?

**Comment by mfeurer** [comment]
Thanks for the heads up @rabsr enjoy your vacation

**Review by mfeurer — COMMENTED** [review]
Thanks a lot for the tests! I just made a PR which makes the examples work. How about I merge this one and you start working on a new PR updating the example for using different metrics?

> If possible, can you highlight what is the use case you are looking forward to?

I'm planning to use it to generate meta-data for the Auto-sklearn 2.0 project.

**Comment by rabsr** [comment]
> How about I merge this one and you start working on a new PR updating the example for using different metrics?

That works for me. You can go ahead with this PR.

**Review by mfeurer — APPROVED** [review]
